R/data.table.R

```diff
@@ -2940,14 +2940,27 @@ isReallyReal <- function(x) {
       if (length(RHS) != nrow(x)) stop("RHS of ", operator, " is length ",length(RHS)," which is not 1 or nrow (",nrow(x),"). For robustness, no recycling is allowed (other than of length 1 RHS). Consider %in% instead.")
       return(NULL) # DT[colA == colB] regular element-wise vector scan
     }
-    if ((is.integer(x[[col]]) && is.double(RHS) && isReallyReal(RHS)) ||
-        (mode(x[[col]]) != mode(RHS) &&
-         !(class(x[[col]]) %in% c("character", "factor") && class(RHS) %in% c("character", "factor"))) ||
-        (is.factor(x[[col]]) && !is.factor(RHS) && mode(RHS)=="numeric") ) { # fringe case, #1361. TODO: cleaner way of doing these checks.
-      # re-direct all non-matching mode cases to base R, as data.table's binary
-      # search based join is strict in types. #957 and #961.
+
+    # fringe cases, #1361.
+    # re-direct all non-matching mode cases to base R, as data.table's binary
+    # search based join is strict in types. #957 and #961.
+    if (is.integer(x[[col]]) && is.double(RHS) && isReallyReal(RHS)) {
+      return(NULL)
+    }
+    if (is.integer(RHS) && is.double(x[[col]]) && isReallyReal(x[[col]])) {
+      return(NULL)
```

Comment:
This section is nice find. Like the solution (although, very minor, I prefer compact lines of code to avoid scrolling).
This particular clause (`is.integer(RHS)) && is.ReallyReal(x[[col]])`) wasn't in the original. Are we sure we want to add it (there isn't any commentary in the PR highlighting this is a change)?  If `x[[col]]` is double and really is real,  matching whole numbers into it is ok and valid isn't it? In the same way as whole numbers stored as double RHS will perform.   Otherwise `isReallyReal()` could be invoked every time on a potentially very long column in `x`.